frontend/src/components/PetitionCard.tsx

```diff
@@ -0,0 +1,244 @@
+      setState(State.Existed);
+    } else {
+      setState(State.Success);
+    }
+  } catch {
+    setState(State.Error);
+  }
+}
+
+function renderInputForQuestion(
+  q: Question & { state: [string, Dispatch<SetStateAction<string>>] }
+) {
+  if ("choices" in q) {
+    const emptyOption = <option key="empty" value=""></option>;
+    return (
+      <select
+        className="w-full h-8 rounded-md text-black p-2"
+        onChange={(e) => q.state[1](e.target.value)}
+        required={q.required ? true : undefined}
+      >
+        {emptyOption}
+        {q.choices.map((c, i) => {
+          return (
+            <option key={i.toString()} value={c}>
+              {c}
+            </option>
+          );
+        })}
+      </select>
+    );
+  }
+  return (
+    <input
+      className="w-full h-8 rounded-md text-black p-2"
+      value={q.state[0]}
+      onInput={(e) => q.state[1](e.target.value)}
+      required={q.required ? true : undefined}
+    ></input>
+  );
+}
+
+function renderExtraQuestions(
+  questions: Array<
+    Question & { state: [string, Dispatch<SetStateAction<string>>] }
+  >
+) {
+  return questions.map((q, i) => (
+    <>
+      <div key={`label-${i.toString()}`}>
+        <label>{q.question}</label>
+      </div>
+      <div key={`input-${i.toString()}`}>{renderInputForQuestion(q)}</div>
+    </>
+  ));
+}
+
+export default function PetitionCard(props: Props) {
+  const petition = props.petition;
+  const [state, setState] = useState(State.Initial);
+  useEffect(() => {
+    if (state != State.Success && ensureSessionStorage()[petition.slug]) {
+      setState(State.Existed);
+    }
+  });
+
+  const extraQuestions = (petition.extraQuestions || []).map((q) => ({
+    state: useState(""),
+    ...q,
+  }));
+
+  const nameId = "name-field"; // CHANGEME: update to useId in react 18
+  const emailId = "email-field"; // CHANGEME: update to useId in react 18
+
+  const [nameInput, setNameInput] = useState("");
+  const [emailInput, setEmailInput] = useState("");
+
+  if (state == State.Submitting) {
+    return (
+      <RoundedBox className={props.className}>
+        <div className="self-center py-20">
+          <FaSpinner className="animate-spin" />
+        </div>
+      </RoundedBox>
+    );
+  }
+  if (state == State.Error) {
+    return (
+      <RoundedBox className={props.className}>
+        <div className="py-20">Er is iets mis gegaan bij het opsturen.</div>
+      </RoundedBox>
+    );
+  }
+  if (state == State.Existed) {
+    return (
+      <RoundedBox className={props.className}>
+        <div className="py-20">Je hebt deze petitie al een keer ingevuld.</div>
+      </RoundedBox>
+    );
+  }
+  if (state == State.Success) {
+    return (
+      <RoundedBox className={props.className}>
+        <div className="py-20">
+          Succes! Check je email om je handtekening definitief te maken.
+        </div>
+      </RoundedBox>
+    );
+  }
+
+  const onSubmitCallback = useCallback(
```

Comment:
handleSubmit* is de typische manier hoe je dit soort dingen noemt (on<Event> -> handle<Event>)